sdk/core/azure-core-amqp/src/impl/rust_amqp/amqp/message_receiver.cpp

```diff
@@ -67,27 +119,76 @@ namespace Azure { namespace Core { namespace Amqp { namespace _detail {
       Azure::Core::_internal::AzureNoReturnPath(
           "MessageReceiverImpl is being destroyed while open.");
     }
-    if (m_link)
-    {
-      m_link.reset();
-    }
-    m_messageQueue.Clear();
   }
 
+  // Nullable<uint32_t> InitialDeliveryCount;
+
+  // Nullable<uint64_t> MaxMessageSize;
+  // uint32_t MaxLinkCredit{};
+
   void MessageReceiverImpl::Open(Azure::Core::Context const& context)
   {
     if (m_options.AuthenticationRequired)
     {
       m_session->GetConnection()->AuthenticateAudience(
           m_session, static_cast<std::string>(m_source.GetAddress()), context);
     }
-    throw std::runtime_error("Not yet implemented");
-  }
 
+    UniqueMessageReceiverOptions options{amqpmessagereceiveroptions_create()};
+
+    InvokeAmqpApi(amqpmessagereceiveroptions_set_name, options, m_options.Name.c_str());
+
+    RustReceiverSettleMode settleMode;
+    switch (m_options.SettleMode)
+    {
+      case ReceiverSettleMode::First:
+        settleMode = RustReceiverSettleMode::First;
+        break;
+      case ReceiverSettleMode::Second:
+        settleMode = RustReceiverSettleMode::Second;
+        break;
+    }
+    InvokeAmqpApi(amqpmessagereceiveroptions_set_receiver_settle_mode, options, settleMode);
+    InvokeAmqpApi(
+        amqpmessagereceiveroptions_set_target,
+        options,
+        Models::_detail::AmqpTargetFactory::ToImplementation(m_options.MessageTarget));
+    InvokeAmqpApi(
+        amqpmessagereceiveroptions_set_properties,
+        options,
+        Models::_detail::AmqpValueFactory::ToImplementation(m_options.Properties.AsAmqpValue()));
+
+    Common::_detail::CallContext callContext(
+        Common::_detail::GlobalStateHolder::GlobalStateInstance()->GetRuntimeContext(), context);
+
+    if (amqpmessagereceiver_attach(
+            callContext.GetCallContext(),
+            m_receiver.get(),
+            m_session->GetAmqpSession().get(),
+            Models::_detail::AmqpSourceFactory::ToImplementation(m_source),
+            options.get()))
+    {
+      throw std::runtime_error("Failed to attach message receiver: " + callContext.GetError());
+    }
+    else
+    {
+      m_receiverOpen = true;
+    }
+  }
   void MessageReceiverImpl::Close(Context const& context)
   {
-    throw std::runtime_error("Not yet implemented");
-    (void)context;
-  }
+    if (m_receiver)
+    {
+      Common::_detail::CallContext callContext(
+          Common::_detail::GlobalStateHolder::GlobalStateInstance()->GetRuntimeContext(), context);
 
+      // Even if the detach fails, we still want to consider the receiver closed.
+      m_receiverOpen = false;
+      if (amqpmessagereceiver_detach_and_release(
+              callContext.GetCallContext(), m_receiver.release()))
+      {
+        throw std::runtime_error("Failed to detach message receiver: " + callContext.GetError());
```

Comment:
Is it expected that the caller will recover from the exception or just that they *might*? Seems odd to support exceptions as control flow as opposed to just having the caller treat the current object state as "corrupt" should an exception be thrown in setup or teardown.

Reply:
This is functionally equivalent to a free or delete failing - there's really no recovery mechanism possible for it. And the primary scenario for this is in an explicit `Close` method rather than in the destructor (Rust equivalent: `drop()` function).

Once the close has failed, there's not much you can do.

In general, the AMQP APIs throw exceptions on things which are likely to be developer caused, and return errors on things that aren't recoverable. This is an example of an unrecoverable error.